### PR TITLE
feat(inventory) change report

### LINKED ISSUE #3975: Add report for all changes to inventory items, not just for a single inventory item
After demonstrating our "Inventory History" feature to the Medical Director of Vanga Hospital, he requested we provide the ability to view a global report of all items that have changed, not the individual history item by item.  This would allow him to view any changes at any time.

### PR DESCRIPTION
**Pull request by jeremielodi**
closes #3975

### CODE REVIEW

**Comment by kwilu** [comment]
Hello @jeremielodi, it seems like you have referenced #3975 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3975.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

File: client/src/modules/reports/generate/inventoryChange/inventoryChange.config.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  'AppCache', 'reportData', '$state',
+];
+
+function InventoryChangeController($sce, Notify, SavedReports, AppCache, reportData, $state) {
+  const vm = this;
+  const cache = new AppCache('inventoryChange');
+  const reportUrl = 'reports/inventory/changes/';
+
+  vm.previewGenerated = false;
+  vm.reportDetails = {
```
Comment (jniles) [inline client/src/modules/reports/generate/inventoryChange/inventoryChange.config.js]: These looks like they are copied from somewhere.  Please remove all the code you didn't write and didn't use.

File: client/src/i18n/en/report.json [hunk 2]
```diff
@@ -106,6 +106,10 @@
       "TITLE": "Debtors summary report",
       "DESCRIPTION": "This report details the amount to paid in every inventory group for an invoice during a period, for a specific debor group"
     },
+    "INVENTORY_CHANGE" : {
+      "TITLE" : "Inventory change report",
```
Comment (jniles) [inline client/src/i18n/en/report.json]: Let's call this "Inventory Changes Report"

File: client/src/modules/reports/generate/inventoryChange/inventoryChange.config.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  vm.onChangeShowRemainDetails = value => {
+    vm.reportDetails.shouldShowRemainDetails = value;
+  };
+
+  vm.onChangeShowPaidDetails = value => {
+    vm.reportDetails.shouldShowPaidDetails = value;
+  };
+
+  vm.onChangeShowInvoicedDetails = value => {
```
Comment (jniles) [inline client/src/modules/reports/generate/inventoryChange/inventoryChange.config.js]: It doesn't look like you use any of this.  Please remove anything you don't use or need.

File: server/controllers/inventory/reports/changes.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
+
+/**
+ * @overview Price Report
```
Comment (jniles) [inline server/controllers/inventory/reports/changes.js]: I don't think this is a price report.  Please update or remove these comments.

File: server/models/migrations/next/migrate.sql [hunk 5]
```diff
@@ -191,3 +191,13 @@ INSERT INTO `survey_form_type` (`id`, `label`, `type`, `is_list`) VALUES
 
 -- BHIMA DATA COLLECTOR
 -- 2019-11-05
+
+-- inventory change report
+--- by jeremielodi
+--- 2019-11-18
+
+INSERT INTO unit VALUES
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: 👍

File: server/controllers/inventory/reports/changes.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  const params = _.clone(req.query);
+
+  const qs = _.extend(req.query, {
+    csvKey : 'groups',
+    footerRight : '[page] / [toPage]',
+    footerFontSize : '7',
+  });
+  const metadata = _.clone(req.session);
+
+  let report;
```
Comment (jniles) [inline server/controllers/inventory/reports/changes.js]: You can move this `let` into a `const` in the `try/catch` block.

File: server/controllers/inventory/reports/changes.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+  try {
+    report = new ReportManager(TEMPLATE, metadata, qs);
+    const { dateFrom, dateTo } = params;
+
+    const inventorySql = `
+      SELECT DISTINCT BUID(iv.uuid) as uuid, iv.text AS inventory_name
```
Comment (jniles) [inline server/controllers/inventory/reports/changes.js:31]: Why do we need the current inventory items?  Can't we just use the data in the logs?

File: server/controllers/inventory/reports/changes.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    const renderResult = await report.render({ inventories, dateFrom, dateTo });
+    res.set(renderResult.headers).send(renderResult.report);
+  } catch (e) {
+    next(e);
+  }
+}
+
+function formatKeys(record) {
+  const removables = ['group_uuid', 'type_id', 'unit_id'];
+  removables.forEach(r => {
```
Comment (jniles) [inline server/controllers/inventory/reports/changes.js]: Just use lodash:
```js
_.omit(record, removables);
```

File: server/controllers/inventory/reports/changes.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    delete record[r];
+  });
+  if (record.text) {
+    record.label = _.clone(record.text);
```
Comment (jniles) [inline server/controllers/inventory/reports/changes.js]: I don't think you need `_.clone` here.  I think you can just set them equivalent.

File: server/controllers/inventory/reports/changes.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    delete record.text;
+  }
+  return record;
```
Comment (jniles) [inline server/controllers/inventory/reports/changes.js]: It looks like your entire `formatKeys` function could be:
```js
function formatKeys(record) {
  record.label = record.text;
  return _.omit(record, ['group_uuid', 'type_id', 'unit_id', 'text']);
}
```

Try it and see if that works.

**Review by jniles — CHANGES_REQUESTED** [review]
Looks like there is quite a few lines copied from other reports that make this PR hard to read and evaluate.  Could you clean those up and ping me back for a review?

**Comment by jniles** [comment]
@jeremielodi what is the status of this PR?

**Comment by jniles** [comment]
@jeremielodi can you make the changes to this PR for it to be merged?

**Comment by kwilu** [comment]
Heads up @jeremielodi, we just merged some commits that conflict with the changes your made in this pull request! You can review this repository's [recent commits](https://github.com/IMA-WorldHealth/bhima/commits/master) to see where the conflicts occur. Please rebase your feature branch against the `upstream/master` branch and [resolve](https://zulip.readthedocs.io/en/latest/git/troubleshooting.html#recover-from-a-git-rebase-failure) your pull request's merge conflicts accordingly.

**Comment by jniles** [comment]
@jeremielodi can you update this PR and fix things according to the review so we can merge?

**Comment by jniles** [comment]
@jeremielodi, it doesn't look like you made any of the changes requested in the review.  Was there a reason for this?

**Review by jniles — APPROVED** [review]

**Comment by kwilu** [comment]
Heads up @jeremielodi, we just merged some commits that conflict with the changes your made in this pull request! You can review this repository's [recent commits](https://github.com/IMA-WorldHealth/bhima/commits/master) to see where the conflicts occur. Please rebase your feature branch against the `upstream/master` branch and [resolve](https://zulip.readthedocs.io/en/latest/git/troubleshooting.html#recover-from-a-git-rebase-failure) your pull request's merge conflicts accordingly.

**Pull request merged**